src/database/migrations/0010-insert-canonical_fields.sql

```diff
@@ -0,0 +1,285 @@
+INSERT INTO canonical_fields (id, label, short_code, data_type) OVERRIDING SYSTEM VALUE VALUES
```

Comment:
@bickelj [wrote](https://github.com/PhilanthropyDataCommons/service/pull/164#issuecomment-1466445071) (and I edited for brevity):

> Another problem arises when someone tries to create application forms in a test instance prior to posting data to the official instance. It is both simpler and easier to clone, build, and run the software and then post an application form with existing known canonical field IDs.

I don't think I fully understand the ID concern! Is that not the purpose of the `shortCode` field? https://github.com/PhilanthropyDataCommons/service/blob/f585cf8712bbc54e3c65247e788abf4a5c98cbe4/src/database/migrations/0001-create-canonical_fields.sql#L12

> ... suppose you start fresh and add canonical fields, and then afterward attempt to add the "official" canonical fields from script and get a collision on `shortCode`. There is no guarantee that your new canonical field matches that canonical field and now you have to go and start from scratch locally again (or do some manual deletes on the database).

This sounds like a question of assigning short codes, and in particular is an artifact of a [current technical limitation](https://github.com/PhilanthropyDataCommons/service/issues/104#issuecomment-1419687999):

> At the fundamental level currently all fields need to have a canonical field underneath them.

Taking that constraint at face value, it sounds like we'd need to either be mixing static and dynamic values, or writing, reviewing, merging, and deploying a database migration each time a partner wanted to import data with a new field. In each case, managing canonical fields by way of database migrations sounds untenable! Am I overstating the problem? Or, what am I missing?

Reply:
> I don't think I fully understand the ID concern! Is that not the purpose of the `shortCode` field?
> 
> https://github.com/PhilanthropyDataCommons/service/blob/f585cf8712bbc54e3c65247e788abf4a5c98cbe4/src/database/migrations/0001-create-canonical_fields.sql#L12

Sure, we can use `shortCode` instead, and go a roundabout way of getting the necessary auto-generated IDs in order to make the post to `/applicationForms`. In other words, we can require "first, get canonical fields, then second post an application form" and that is technically OK but requires more steps. Originally, I think Dan had the `canonicalFields` as a special table in that rather than referencing auto-generated IDs we would join on `shortCode`. In all other places we use the auto-generated IDs. In general, auto-generated IDs are necessarily exposed so that callers can use them in endpoints such as `GET /applicationForm/{id}`. For consistency, I argued that we should use the canonical field auto-generated IDs like any other. Or perhaps we could go another way and use natural keys throughout, i.e. something akin to `shortCode` in every table, but that was an early decision we decided against. All that said, the whole system from top to bottom is up for revision at any time. It is better to get it closer to what is needed sooner than later. I am not stuck on this approach if there is a better one. Perhaps the complexity (from a caller attempting to create an application form perspective) of the approach you described is still better overall. Some of that complexity will be borne by our team (e.g. in the https://github.com/PhilanthropyDataCommons/service/pull/164 repo rather than the actual caller).

> This sounds like a question of assigning short codes, and in particular is an artifact of a [current technical limitation](https://github.com/PhilanthropyDataCommons/service/issues/104#issuecomment-1419687999):
> 
> > At the fundamental level currently all fields need to have a canonical field underneath them.

It is a modeling or design decision that creates a constraint, for sure. Constraints are often a good thing.

> Taking that constraint at face value, it sounds like we'd need to either be mixing static and dynamic values, or writing, reviewing, merging, and deploying a database migration each time a partner wanted to import data with a new field. In each case, managing canonical fields by way of database migrations sounds untenable! Am I overstating the problem? Or, what am I missing?

This is the clearest and best objection to the approach in this PR, yes.

The goal of the PDC as I understand it is to provide a means to curate a list of common application form fields. If the data model is getting in the way, we need a fresh data model. If this approach gets in the way, we need a fresh approach.

In OTS chat you mentioned that the problem is similar to that of tags for large data archives. Tags could be totally free-form which results in an explosion of very specific or duplicated tags. Tags could be totally managed and controlled by a central authority which results in missing or bad tags. Or there could be some middle ground where a group of moderators links or updates free-form tags into the most useful form. I agree this problem is similar here.

Reply:
See below. The team agreed we can use a SQL script off to the side, to be superseded by a script that interacts with the API.